development-practices/defining-the-boundaries-of-a-package.md

```diff
@@ -0,0 +1,107 @@
+
+![Package C with a dependency cycle](../.gitbook/assets/package-dependency.png)
+
+These packages are impossible to deploy on a new Org. One can't deploy Package A without deploying Package C and Package C itself can't be deployed without deploying Package A resulting in a standoff. Hence, cycles should be broken either by inverting the dependencies or by introducing a shared package which shares the common components for both Package A and Package C.
+
+#### The Stable Dependencies Principle (SDP)
+
+> **The dependencies between packages in a design should be in the direction of the stability of the packages. A package should only depend upon packages that are more stable that it is.**
+
+Designs cannot be completely static. Some volatility is necessary if the design is to be maintained. We accomplish this by conforming to the _Common Closure Principle_ (**CCP**). By using this principle, we create packages that are sensitive to certain kinds of changes. These packages are designed to be volatile. We expect them to change.
+
+Any package that we expect to be volatile should not be depended upon by a package that is difficult to change! Otherwise the volatile package will also be difficult to change.
+
+By conforming to the **SDP**, we ensure that modules that are designed to be instable (i.e. easy to change) are not depended upon by modules that are more stable (i.e. harder to change) than they are.
+
+#### The Stable Abstractions Principle (SAP) \[Code-Related]
+
+> **Packages that are maximally stable should be maximally abstract. Instable packages should be concrete. The abstraction of a package should be in proportion to its stability.**
+
+This principle sets up a relationship between stability and abstractness. It says that a stable package should also be abstract so that its stability does not prevent it from being extended. On the other hand, it says that an instable package should be concrete since its instability allows the concrete code within it to be easily changed.
+
+![](../.gitbook/assets/sdp-uml-reference.png)
+
+Consider the “_Copy_” program above. The “_Reader_” and “_Writer_” classes are abstract classes. They are highly stable since they depend upon nothing and are depended upon by “_Copy_” and all their derivatives. Yet, “_Reader_” and “_Writer_” can be extended, without modification, to deal with many kinds of I/O devices.
+
+Thus, if a package is to be stable, it should also consist of abstract classes so that it can be extended. Stable packages that are extensible are flexible and do not constrain the design.
+
+The **SAP** and the **SDP** combined amount to the Dependency Inversion Principle for Packages. This is true because the **SDP** says that dependencies should run in the direction of stability, and the **SAP** says that stability implies abstraction. Thus, dependencies run in the direction of abstraction.
```

Comment:
```suggestion
The **SAP** and the **SDP** combined amount to the Dependency Inversion technique for Packages. This is true because the **SDP** says that dependencies should run in the direction of stability, and the **SAP** says that stability implies abstraction. Thus, dependencies run in the direction of abstraction.
```